packages/controllers/src/permissions/Caveat.ts

```diff
@@ -0,0 +1,257 @@
+   * created or mutated.
+   *
+   * The validator should throw an appropriate JSON-RPC error if validation fails.
+   *
+   * If no validator is specified, no validation of caveat values will be
+   * performed. Although caveats can also be validated by permission validators,
+   * validating caveat values separately is strongly recommended.
+   */
+  validator?: CaveatValidator<SpecifiedCaveat>;
+};
+
+/**
+ * TODO:docs
+ */
+export type CaveatSpecificationMap<
+  Specification extends CaveatSpecificationConstraint,
+> = {
+  [Key in Specification['type']]: Specification extends CaveatSpecification<
+    Caveat<Key, any>
+  >
+    ? Specification
+    : never;
+};
+
+/**
+ * TODO:docs
+ */
+export type ExtractCaveats<
+  Specification extends CaveatSpecification<CaveatConstraint>,
+> = Specification extends any
+  ? Caveat<
+      Specification['type'],
+      ExtractCaveatValueFromDecorator<Specification['decorator']>
+    >
+  : never;
+
+/**
+ * Internal utility type, because using parameterized types in conditional types
+ * causes weird things to happen.
+ */
+type _ExtractCaveat<
+  CaveatUnion extends CaveatConstraint,
+  CaveatType extends string,
+> = CaveatUnion extends Caveat<CaveatType, Json> ? CaveatUnion : never;
+
+/**
+ * Extracts the type of a specific {@link Caveat} from a union of caveat
+ * specifications.
+ *
+ * @template CaveatSpecifications - The union of all caveat specifications.
+ * @template CaveatType - The type of the caveat to extract.
+ */
+export type ExtractCaveat<
+  CaveatSpecifications extends CaveatSpecification<CaveatConstraint>,
+  CaveatType extends string,
+> = _ExtractCaveat<ExtractCaveats<CaveatSpecifications>, CaveatType>;
+
+/**
+ * Extracts the value type of a specific {@link Caveat} from a union of caveat
+ * specifications.
+ *
+ * @template CaveatSpecifications - The union of all caveat specifications.
+ * @template CaveatType - The type of the caveat whose value to extract.
+ */
+export type ExtractCaveatValue<
+  CaveatSpecifications extends CaveatSpecification<CaveatConstraint>,
+  CaveatType extends string,
+> = ExtractCaveat<CaveatSpecifications, CaveatType>['value'];
+
+/**
+ * Decorate a restricted method implementation with its caveats.
+ *
+ * Note that all caveat functions (i.e. the argument and return value of the
+ * decorator) must be awaited.
+ */
+export function decorateWithCaveats<
+  CaveatSpecifications extends CaveatSpecification<CaveatConstraint>,
+>(
+  methodImplementation: RestrictedMethod<RestrictedMethodParameters, Json>,
+  permission: Readonly<PermissionConstraint>, // bound to the requesting origin
+  caveatSpecifications: CaveatSpecificationMap<CaveatSpecifications>, // all caveat implementations
+): RestrictedMethod<RestrictedMethodParameters, Json> {
+  const { caveats } = permission;
+  if (!caveats) {
+    return methodImplementation;
+  }
+
+  let decorated = methodImplementation as AsyncRestrictedMethod<
+    RestrictedMethodParameters,
+    Json
+  >;
```

Comment:
This cast hides a potentially important change in behaviour, which is how errors are propagated. A synchronous method would throw, whereas an async method would reject.

Rather than casting, we could wrap this in an async method instead:

```suggestion
  let decorated = async (
    args: Parameters<RestrictedMethod<RestrictedMethodParameters, Json>>[0],
  ) => methodImplementation(args);
```

This ensures the error handling behavior is normalized, and it lets us drop the cast.

Reply:
This suggestion also broke unit tests FYI. The unit test for this method wrongly assumed that the method passed in was synchronous.